proposal: console changes

maggu2810 opened the pull request:
NOTE: This intention of this PR is to give a place for a discussion. Nothing that will be merged (ATM).

First commit:
I split the console bundle in two parts. The one with the interface definition etc. and the one that contains the implementation to support the Eclipse Equinox OSGi framework (see first commit).

Second commit:
Add a simple bundle that adds support for the karaf shell. I call it simple, because it provides only one karaf command (scope: esh, command name: smarthome), that takes the current commands as arguments.
If a command name is unique in a set of scopes and command names, you could omit the scope. So instead of calling <pre>esh:smarthome things list</pre> you could also call <pre>smarthome things list</pre> if there is no smarthome command in another scope.

maggu2810 commented:
The ConsoleActivator creates two service trackers. One for the ItemRegistry and one for the EventPublisher. Both are not used at the moment. What is the intention of this trackers, if they are not used. Have you created the service tracker for the extension of the equinox console to use them?

maggu2810 commented:
I changed the commits and added a new one.
The first and third commit does not depend on Karaf.
That commits should improve the console extension interface / current implementation.
The first one introduce a new interface and implementation, the third one changed the discovery bundle to use the new one.

dnobel commented:
Thanks for the PR. We will take a look at it soon.

maggu2810 commented:
Will add a cleaned up version soon, if you could wait a few minutes...

maggu2810 commented:
Okay, you could start looking at it now.

Changes:
* split bundle io.console in two bundles io.console and io.console.eclipse
* io.console: console interface and base implementation
* io.console.eclipse: implementation for Eclipse Equinox integration
* a command extension should handle exact one command (could still handle a non limited number of subcommands) and a description
* use a standard format for the command extension usage string
* change existing command extensions to fit new implementation


Perhaps for OH2, not for ESH (could remove it, if the rest is accepted and gets to be squashed):

At least I added a simple bundle that adds support for the karaf shell.
I call it simple, because it provides only one karaf command (scope: esh, command name: smarthome), that takes the current commands as arguments.
This is the same way as the current bundle for Eclipse Equinox is implemented (with missing scope).
For Karaf, if a command name is unique in a set of scopes and command names, you could omit the scope. So instead of calling
<pre>esh:smarthome things list</pre>
you could also call
<pre>smarthome things list</pre>
if there is no smarthome command in another scope.

kaikreuzer commented:
I have just had a closer look at the whole topic.
I wonder if we should at all keep our own infrastructure at all - there is an official OSGi recommendation, see http://felix.apache.org/site/rfc-147-overview.html. Although rfc-147 did not make it into the official OSGi R4.2 spec, the way how commands are registered as a service is kept and fully supported by Gogo.
There is not really the need for having org.eclipse.equinox.console - instead, we can by default use the gogo shell directly on Equinox. IF somebody wants to support org.eclipse.equinox.console, we could read the services and make them available just as it is done today.
So effectively, this would mean replacing our ConsoleCommandExtension with a service according to RfC-147. This should also directly work on Karaf, I would assume.

As a side note: I think "smarthome" should be the scope and not a command - the above suggestion would take care of this as well.

maggu2810 commented:
You have already pointed me to that RFC.
I split the bundles so we get a org.eclipse.smarthome.io.console.eclipse, that is using org.eclipse.equinox.console.
I create a simple org.eclipse.smarthome.io.console.karaf that could be used in Karaf.
I would also try to create a org.eclipse.smarthome.io.console.gogo / org.eclipse.smarthome.io.console.rfc147 / or something similar that could be used for the gogo console implementation.
As long as there is no official support in a OSGi spec, we should not replace our ConsoleCommandExtension (IMHO).
Karaf is using another mechanism for the console command extension. I could recheck, but the gogo mechanism is not working using Karaf.

Yes, smarthome should be the scope and the command should be that one of the new ConsoleCommandExtension (getCommand()). This could be handled if this is possible for the choosen implementation.
Karaf checks the bundle manifest if there is a "Karaf-Commands" statement.
If there is one, it inspect all class files and check the annotations in that files.
That is the API, all other stuff is private implementation and not exported.
So I used "smarthome" as command (that is made public by implement Action and added Annotations) and removed it in the evaluation again (=> cmd = args[0], args[0]...args[x-1] = args[1] ...args[x]).
But that is part of console.karaf and perhaps out of scope.

I think we could keep this cleaned up implementation, so a ConsoleCommandExtension has a command, a short desciption and a usage (that could explain sub commands, etc.).

maggu2810 commented:
As long as the rfc-147 is not part of the official spec., we should keep the ConsoleCommandExtension.
We should check, that a simple wrapper could use that classes and create a command extension for a rfc-147 complient console (at runtime).
We should also check if the ConsoleCommandExtension contains all stuff to implement a wrapper for Karaf, etc.

kaikreuzer commented:
> As long as the rfc-147 is not part of the official spec., we should keep the ConsoleCommandExtension.

It will never be part of the official spec, see http://wiki.osgi.org/wiki/RFC_147:
_"OSGi has ceased work on RFC 147 and encouraging people to put their efforts behind Apache Felix Gogo."_
This approach seems to be adopted everywhere. You can find it in Knopflerfish (https://www.knopflerfish.org/releases/3.5/docs/javadoc/org/osgi/service/command/CommandProcessor.html), BndTools (http://bndtools.org/tutorial.html#write-a-command-component) and of course in Felix Gogo, so you can consider it a de-facto standard.
And even in Karaf 4, I am not sure if the "old" Karaf extension mechanism is still recommended - the classes are marked deprecated (https://github.com/apache/karaf/commit/66c52d19b0b779a1c54cb43ffeb342d144cce7e4) and there was a discussion about concentrating on Gogo commands (https://issues.apache.org/jira/browse/KARAF-2772?focusedCommentId=13912730&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13912730)

I therefore do not see any reason why we should behave incompatible to this recommendation.

maggu2810 commented:
> And even in Karaf 4, I am not sure if the "old" Karaf extension mechanism is still recommended

Then the current documentation if wrong, as it is using the new API.
http://repo1.maven.org/maven2/org/apache/karaf/manual/4.0.0.M1/manual-4.0.0.M1.html#Extending

Will have a further look at RFC-147.

Sure, we could remove (an perhaps not necessary) abstraction layer, but I think we have to check the pro and cons.

cons: "some" overhead

I like the abstraction between a ESH specific console command extension and "normal" one.
If I have to create a command for the ESH system, I do it for the ESH framework and not for a specific OSGi framework, RFC, ...

use case: We want to create a UI that shows all commands that extend the "ESH console". With the current solution we could "collect" all implementation of the interface and show them to the user. We could only show that commands and no other ones.
How could we differ other command extensions between them of ESH (need we to do that?)?
We could use the scope, but also we could just say: you should use the scope xyz for ESH console extensions. If we want to change that scope all extensions have to be changed...

If the ESH system would like that all their command extensions offer a function "getInformationWeNeed()" than we could extend the interface and all commands need to implement them.

If someone use an implementation that does not support that RFC he have to create a wrapper. So he needs to be informed for added / removed commands. Can we get a notification if a service adds function to the console using the RFC method?

So, I see no real enhancement to remove that abstraction (as long as I do not implement some tests). Perhaps the memory footprint, system load overhead is a "no go", so we should move to "gogo". ;-)

eclipsewebmaster commented:
:no_entry_sign: 
#### Eclipse validation results:
The pull request did not pass Eclipse validation.
The following users have invalid Signed-off-by footers. Please use the Signed-Off-By (git commit -s) mechanism to <a href="www.eclipse.org/legal/CoO.php">confirm the origin of the submission</a>. The identity in the Signed-off-By must match the Author or the Committer.maggu2810@gmail.com

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

kaikreuzer commented:
> How could we differ other command extensions between them of ESH (need we to do that?)?

Yes, the scope would be the natural identifier for this.

> Can we get a notification if a service adds function to the console using the RFC method?

Well, yes, the command services would simply come and go and you get OSGi events for that.

> So, I see no real enhancement to remove that abstraction (as long as I do not implement some tests). > Perhaps the memory footprint, system load overhead is a "no go", so we should move to "gogo". ;-)

This should be negligible. So ok, as the situation with Karaf is a bit unclear and we might need a wrapper for Karaf after all, I am ok with keeping our own abstraction here. What I would like to make sure though is, that the refactoring allows a Karaf-wrapper with "smarthome" as a scope and not a command. I am not sure, if I fully got the answer to this yet.
And regarding the question whether the Karaf wrapper belongs in ESH or OH2: I would prefer to have it in ESH directly, if we get the dependency managed.

maggu2810 commented:
> I would prefer to have it in ESH directly, if we get the dependency managed.
We could keep the dependency handling in the pom file only, if it is not part of the main bundle. Couldn't we?

* On my system a bundle that is using the RFC 147 implementation does not work out of the box using Karaf.
* The annotations Karaf is now using have to be added on development, so not solution.
* Till now I does not find a proper solution for Karaf to add commands programmatically.

maggu2810 commented:
Unfortunately I have little free time to deal with this (Karaf) problem.
But this is nothing that needs to be solved this week / month (or I am wrong)?

```java
/**
 * The <code>Manager</code> service can be used to programmatically
 * register {@link org.apache.karaf.shell.api.action.Action}s or
 * {@link org.apache.karaf.shell.api.console.Completer}s.
 *
 * Registered objects must be annotated with the {@link Service} annotation.
 *
 * Objects will be registered in the {@link org.apache.karaf.shell.api.console.Registry}
 * associated with this <code>Manager</code>.
 *
 * @see org.apache.karaf.shell.api.console.Registry
 * @see org.apache.karaf.shell.api.action.lifecycle.Service
 */
```

I keep reading...

kaikreuzer commented:
> But this is nothing that needs to be solved this week / month (or I am wrong)?

No, not really. I only want to make sure that this PR is a step in the right direction. I would not like to see us discover that the refactoring would have needed to be done in a completely different way in order to also be able to cover Karaf (in a few weeks).

maggu2810 commented:
This is also my intention. I would like to keep that PR open until this is solved using Karaf or we know, that this is not possible.
Currently I read the Karaf sources and perhaps I am on the right direction.
Will report tomorrow if this is working.

kaikreuzer commented:
ok, I am fine with leaving this open. No hurry at all!

dnobel commented on bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/extensions/GenericConsoleCommandExtension.java:
```diff
@@ -0,0 +1,80 @@
+ * @author Markus Rathgeb - Initial contribution and API
+ *
+ */
+public abstract class GenericConsoleCommandExtension implements ConsoleCommandExtension {
```
One comment from my site: The name GenericConsoleCommandExtension is a little bit confusing to me. It looks like an abstract base class for concrete console command extensions (and you also documented it like this in the class java doc). "Generic" is used in other ESH bundles for concrete service implementations (like the GenericItemProvider). We typically use Base or Abstract as prefix for abstract base classes. So "AbstractConsoleCommandExtension" would be a better name.

maggu2810 replied:
Thanks, will change that.
Generic is for me something that is not specific (so nothing that is concrete IMHO), but Abstract or Base is better, you're right.

dnobel commented on bundles/io/org.eclipse.smarthome.io.console.rfc147/src/main/java/org/eclipse/smarthome/io/console/rfc147/internal/CommandHelp.java:
```diff
@@ -0,0 +1,44 @@
+ * @author Markus Rathgeb - Initial contribution and API
+ *
+ */
+public class CommandHelp extends GenericConsoleCommandExtension {
```
Why not also name it "HelpConsoleCommandExtension" to be aligned with the other extensions?

maggu2810 replied:
You're right (again) ;-)

maggu2810 commented:
Karaf: It is working

```sh
karaf@root()> help
...
smarthome                         Enter the subshell                                                                                                                                                                                                                             
smarthome:items                   Get information about items.                                                                                                                                                                                                                   
smarthome:send                    Send a command to an item.                                                                                                                                                                                                                     
smarthome:status                  Get the current status of an item.                                                                                                                                                                                                             
smarthome:update                  Send a state update to an item.           
...                                                                                                                                                                                                     
karaf@root()> smarthome:items
All (Type=GroupItem, Members=0, State=Uninitialized, Label=All, Category=null, Tags=[home-group])
karaf@root()> smarthome:status All
Uninitialized
```

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

kaikreuzer commented:
> Karaf: It is working

Cool, congrats!

maggu2810 commented:
Are there any comments about the interface versioning?
If I am correct, all ESH bundles are using the same version (0.8.0.qualifier ATM).
As I am still a beginner using the OSGi framework I would like to ask if this is the best practice.

My bundle to integrate the console extensions into Karaf is using the ConsoleCommandExtension and needs the methods the interface in this current PR is using.
See: https://github.com/maggu2810/smarthome/commit/fd6aa84979e792fe5216fb1cff15c5fdd24706c9#diff-a1784761984c28e2d615a0f52064cba1R95
If there is another ConsoleCommandExtension set / added (by DS support) and a method is accessed that is not available (because the ConsoleCommandExtension is too old) the bundle stops working. I could not control if someone outside ESH implements against an old ConsoleCommandExtension interface and add this bundle to his environment. So I have to refer to a special version range (??).
Which version I should depend on?
How do you update all that interface version references?

kaikreuzer commented:
In general, all ESH bundles have the same version - currently 0.8.0.<timestamp>. External bundles that use APIs must be compatible to the API of the used ESH version. This of course means that in theory the ESH APIs must not change from one build to another. And even a minor version change should only bring new stuff, but not break existing things. This is definitely what we aim for, once the project is mature (i.e. has reached version 1.0.0). Currently, we are in incubation mode and thus I do not want to handle it that strict, since it is normal that we are still heavily working on things. This means that whenever a new stable build is released (https://hudson.eclipse.org/smarthome/job/SmartHomeDistributionPublishStable/), all "users" need to be aware of breaking changes and are required to adapt their code. Exactly this must happen for console users after we have merged this PR.
Once we have reached 1.0.0, the project will adhere to the semantic versioning approach (http://semver.org/).
Does this work for you and answers your question?

maggu2810 commented:
All fine for now (till v1.0.0 is released).
So, what should we do next.

I could see here three changes:

split interface / base implementation and the Equinox console + change interface and base implementation
* console{,fw.eclipse}: split current bundle, improve interface …
* console.cmd.discovery: use new interface …
* console.cmd.scriptengine: use new interface …

add support for rfc147
* console.fw.rfc147: Support for RFC 147 …

add support for Karaf
* console.fw.karaf: add simple implementation for Karaf …

* the first one seems ok for me
* for rfc147 I am not using the public string constants so do not depend on Apache GoGo bundles (see Rfc147Manager)
* for Karaf support we need at least one additional bundle in Orbit or something similar

So, how could we proceed?
I could split stuff and use separate PR if adding a new bundle to Orbit is a long time process.

What are the requirements for using external bundles? Is it correct that the IP team have to inspect the bundles etc. so we cannot use some external p2 repo?

kaikreuzer commented:
> I could see here three changes

Yes, this looks good. The second one (rfc147) is optional as nobody really needs it right now. But as you have more or less already completely done it, it is a nice addition for sure.

> I could split stuff and use separate PR if adding a new bundle to Orbit is a long time process.

Probably not a bad idea. Just take out the Karaf bundle from this PR and we will do this separately.

> What are the requirements for using external bundles? Is it correct that the IP team have to inspect the bundles etc. so we cannot use some external p2 repo?

Yes, I need an approval (though an CQ for a works-with dependency, see a) ii. in https://eclipse.org/org/documents/Eclipse_Policy_and_Procedure_for_3rd_Party_Dependencies_Final.pdf) As it is a library from Apache, the approval will be simply. Nonetheless, we need to get it in some p2 repo and adding it to Orbit might take a while. So the simpler option is probably to set up an external p2 repo for now. I will take care of this at BinTray, where I am already hosting the EnOcean deps as well (see https://bintray.com/kaikreuzer/esh-deps).
Could you please give me the link (again?) to the exact jar file that we need?

maggu2810 commented:
You could download the jar from Maven Central:
http://search.maven.org/#artifactdetails%7Corg.apache.karaf.shell%7Corg.apache.karaf.shell.core%7C4.0.0.M2%7Cbundle

Or you could download the release from an Apache Mirror:
https://www.apache.org/dyn/closer.cgi/karaf/4.0.0.M2/apache-karaf-4.0.0.M2.tar.gz
The jar is then placed in the archive using that path: 'apache-karaf-4.0.0.M2/system/org/apache/karaf/shell/org.apache.karaf.shell.core/4.0.0.M2/org.apache.karaf.shell.core-4.0.0.M2.jar'

maggu2810 commented:
Created a new one with the squahed commits (without Karaf): #237 
Will create a new one (and close this one) for Karaf if #237 gets merged.

kaikreuzer commented:
Thanks @maggu2810!
I have now created a p2 repository at http://dl.bintray.com/kaikreuzer/esh-deps-p2 which contains the karaf shell bundle. Since the repo does not contain any feature though, I am not sure, if it is usable in the target platform - please have a try and let me know.
I have also requested approval to use this bundle as a works-with dependency, so that the way is clear for the Karaf PR.

kaikreuzer commented:
Here's the link to the request: https://dev.eclipse.org/mhonarc/lists/iot-pmc/msg00375.html
I close this PR now as your proposal has been discussed successfully.

kaikreuzer commented:
@maggu2810 The Karaf shell bundle has been approved as a works-with dependency, so the way is clear to create a PR for it.
I didn't yet succeed setting up a p2 repo at BinTray for it, I will do so over the weekend and include this in the ESH TP, so that all compile time dependencies should then be available.

maggu2810 commented:
Using a p2 repo created with package-drone worked for me. Sorry, but have no time at the moment to play with BinTray at the moment.
Will create a PR next week for the karaf bundle.

maggu2810 commented:
I have created PR #247
